src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,563 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+
+namespace osp
+{
+
+#ifdef __GNUC__
+
+    inline int ctz(uint64_t a) noexcept { return __builtin_ctzll(a); }
+
+#elif defined(_MSC_VER)
+
+    #include <intrin.h>
+
+    inline int ctz(uint64_t a) noexcept
+    {
+        unsigned long int b = 0;
+        _BitScanForward64(&b, a);
+        return b;
+    }
+#elif
+
+    static_assert(false, "Missing ctz 'count trailing zeros' implementation for this compiler");
+
+#endif
+
+/**
+ * @brief Divide and round up
+ */
+template<typename NUM_T, typename DENOM_T>
+constexpr decltype(auto) div_ceil(NUM_T num, DENOM_T denom) noexcept
```

Comment:
I noticed in the recent C++23 standards committee mailing there was a paper about making most floating point functions *NOT* constexpr, to prevent the compiler from implementing the math differently than what the target platform's CPU would do.

Not sure it matters, but something to be aware of.

Reply:
This function is for integers only, I'll make sure of that. Though from dealing with some of the coordinate space and precision stuff, I can see why that would make sense.